File: backend/routes/parameters.ts
import express from 'express';
import { parameterService } from '../services/parameter-service.ts';
import { validateUUID } from '../utils/validation.ts';

const router = express.Router();

// Middleware to check authentication from Bearer token
async function requireAuth(req: express.Request, res: express.Response, next: express.NextFunction) {
  try {
    const token = req.headers.authorization?.replace('Bearer ', '');
    if (!token) {
      return res.status(401).json({ success: false, error: 'Authentication required' });
    }

    const { sessionRepository } = await import('../db/repositories/session-repository.ts');
    const session = await sessionRepository.findByToken(token);

    if (!session) {
      return res.status(401).json({ success: false, error: 'Invalid or expired session' });
    }

    const { userService } = await import('../services/user-service.ts');
    const user = await userService.getUserById(session.user_id);
    if (!user) {
      return res.status(401).json({ success: false, error: 'User not found' });
    }

    (req as any).user = user;
    next();
  } catch (error) {
    console.error('Auth error:', error);
    return res.status(401).json({ success: false, error: 'Authentication failed' });
  }
}

// Middleware to check admin role
function requireAdmin(req: express.Request, res: express.Response, next: express.NextFunction) {
  const user = (req as any).user;
  if (!user || (user.role !== 'super_admin' && user.role !== 'company_admin')) {
    return res.status(403).json({ success: false, error: 'Admin access required' });
  }
  next();
}

// GET /api/parameters - Get all parameters or filtered by category/type
router.get('/', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const categoryId = req.query.categoryId as string;
    const type = req.query.type as string;

    let parameters;

    // Filter by category or type
    if (categoryId) {
      parameters = await parameterService.getParametersByCategory(categoryId);
    } else if (type) {
      parameters = await parameterService.getParametersByType(type as any);
    } else {
      parameters = await parameterService.getAllParameters();
    }

    return res.json({ success: true, parameters });
  } catch (error) {
    console.error('Get parameters error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get parameters'
    });
  }
});

// POST /api/parameters - Create new parameter (admin only)
router.post('/', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const { name, type, categoryId, value, description, metadata } = req.body;

    // Validate required fields
    if (!name || !type || !categoryId) {
      return res.status(400).json({
        success: false,
        error: 'Missing required fields: name, type, categoryId'
      });
    }

    // Validate type
    const validTypes = ['structured', 'narrative', 'guardrails'];
    if (!validTypes.includes(type)) {
      return res.status(400).json({
        success: false,
        error: `Invalid type. Must be one of: ${validTypes.join(', ')}`
      });
    }

    // Create parameter
    const parameter = await parameterService.createParameter({
      name,
      type,
      categoryId,
      value: value || null,
      description: description || null,
      metadata: metadata || null,
    });

    if (!parameter) {
      return res.status(400).json({ success: false, error: 'Failed to create parameter' });
    }

    return res.status(201).json({ success: true, parameter });
  } catch (error) {
    console.error('Create parameter error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to create parameter'
    });
  }
});

// POST /api/parameters/reset - Reset parameters to defaults (admin only)
router.post('/reset', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    // Reset parameters to defaults
    await parameterService.resetToDefaults();

    return res.json({
      success: true,
      message: 'Parameters reset to defaults successfully'
    });
  } catch (error) {
    console.error('Reset parameters error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to reset parameters'
    });
  }
});

// GET /api/parameters/:id - Get parameter by ID
router.get('/:id', requireAuth, async (req: express.Request, res: express.Response) => {
  try {
    const parameterId = req.params.id;

    // Validate UUID
    if (!validateUUID(parameterId)) {
      return res.status(400).json({ success: false, error: 'Invalid parameter ID format' });
    }

    // Get parameter
    const parameter = await parameterService.getParameterById(parameterId);

    if (!parameter) {
      return res.status(404).json({ success: false, error: 'Parameter not found' });
    }

    return res.json({ success: true, parameter });
  } catch (error) {
    console.error('Get parameter error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to get parameter'
    });
  }
});

// PATCH /api/parameters/:id - Update parameter (admin only)
router.patch('/:id', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const parameterId = req.params.id;

    // Validate UUID
    if (!validateUUID(parameterId)) {
      return res.status(400).json({ success: false, error: 'Invalid parameter ID format' });
    }

    const { name, type, value, description, metadata } = req.body;

    // Validate type if provided
    if (type) {
      const validTypes = ['structured', 'narrative', 'guardrails'];
      if (!validTypes.includes(type)) {
        return res.status(400).json({
          success: false,
          error: `Invalid type. Must be one of: ${validTypes.join(', ')}`
        });
      }
    }

    // Update parameter
    const updatedParameter = await parameterService.updateParameter(parameterId, {
      name,
      type,
      value,
      description,
      metadata,
    });

    if (!updatedParameter) {
      return res.status(404).json({ success: false, error: 'Parameter not found' });
    }

    return res.json({ success: true, parameter: updatedParameter });
  } catch (error) {
    console.error('Update parameter error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to update parameter'
    });
  }
});

// DELETE /api/parameters/:id - Delete parameter (admin only)
router.delete('/:id', requireAuth, requireAdmin, async (req: express.Request, res: express.Response) => {
  try {
    const parameterId = req.params.id;

    // Validate UUID
    if (!validateUUID(parameterId)) {
      return res.status(400).json({ success: false, error: 'Invalid parameter ID format' });
    }

    // Delete parameter
    const success = await parameterService.deleteParameter(parameterId);

    if (!success) {
      return res.status(404).json({ success: false, error: 'Parameter not found' });
    }

    return res.json({ success: true, message: 'Parameter deleted successfully' });
  } catch (error) {
    console.error('Delete parameter error:', error);
    return res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Failed to delete parameter'
    });
  }
});

export default router;

File: backend-python/app/routers/parameters.py
"""
Parameters Router
Parameter management endpoints
"""

import logging
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, status, Depends, Body

from app.services.parameter_service import parameter_service
from app.middleware.auth import get_current_user, require_admin
from app.models.user import UserData
from app.utils.validation import validate_uuid

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/parameters", tags=["Parameters"])


@router.get("")
async def get_parameters(
    user: Annotated[UserData, Depends(get_current_user)],
    categoryId: str | None = None,
    type: str | None = None,
):
    """Get all parameters or filtered by category/type."""
    try:
        if categoryId:
            parameters = await parameter_service.get_parameters_by_category(categoryId)
        elif type:
            parameters = await parameter_service.get_parameters_by_type(type)
        else:
            parameters = await parameter_service.get_all_parameters()

        return {"success": True, "parameters": [p.model_dump() for p in parameters]}
    except Exception as e:
        logger.error(f"Get parameters error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_parameter(
    admin: Annotated[UserData, Depends(require_admin)],
    name: str = Body(...),
    type: str = Body(...),
    categoryId: str = Body(...),
    value: Any | None = Body(None),
    description: str | None = Body(None),
    metadata: dict[str, Any] | None = Body(None),
):
    """Create new parameter (admin only)."""
    try:
        # Validate required fields
        if not name or not type or not categoryId:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Missing required fields: name, type, categoryId",
            )

        # Validate type
        valid_types = ["structured", "narrative", "guardrails"]
        if type not in valid_types:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid type. Must be one of: {', '.join(valid_types)}",
            )

        parameter = await parameter_service.create_parameter({
            "name": name,
            "type": type,
            "category_id": categoryId,
            "value": value,
            "description": description,
            "metadata": metadata,
        })

        if not parameter:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Failed to create parameter")

        return {"success": True, "parameter": parameter.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Create parameter error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("/reset")
async def reset_parameters(
    admin: Annotated[UserData, Depends(require_admin)],
):
    """Reset parameters to defaults (admin only)."""
    try:
        await parameter_service.reset_to_defaults()

        return {"success": True, "message": "Parameters reset to defaults successfully"}
    except Exception as e:
        logger.error(f"Reset parameters error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/{parameter_id}")
async def get_parameter(
    parameter_id: str,
    user: Annotated[UserData, Depends(get_current_user)],
):
    """Get parameter by ID."""
    try:
        if not validate_uuid(parameter_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid parameter ID format")

        parameter = await parameter_service.get_parameter_by_id(parameter_id)

        if not parameter:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Parameter not found")

        return {"success": True, "parameter": parameter.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get parameter error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.patch("/{parameter_id}")
async def update_parameter(
    parameter_id: str,
    admin: Annotated[UserData, Depends(require_admin)],
    name: str | None = Body(None),
    type: str | None = Body(None),
    value: Any | None = Body(None),
    description: str | None = Body(None),
    metadata: dict[str, Any] | None = Body(None),
):
    """Update parameter (admin only)."""
    try:
        if not validate_uuid(parameter_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid parameter ID format")

        # Validate type if provided
        if type:
            valid_types = ["structured", "narrative", "guardrails"]
            if type not in valid_types:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Invalid type. Must be one of: {', '.join(valid_types)}",
                )

        update_data: dict[str, Any] = {}
        if name is not None:
            update_data["name"] = name
        if type is not None:
            update_data["type"] = type
        if value is not None:
            update_data["value"] = value
        if description is not None:
            update_data["description"] = description
        if metadata is not None:
            update_data["metadata"] = metadata

        updated_parameter = await parameter_service.update_parameter(parameter_id, update_data)

        if not updated_parameter:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Parameter not found")

        return {"success": True, "parameter": updated_parameter.model_dump()}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Update parameter error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.delete("/{parameter_id}")
async def delete_parameter(
    parameter_id: str,
    admin: Annotated[UserData, Depends(require_admin)],
):
    """Delete parameter (admin only)."""
    try:
        if not validate_uuid(parameter_id):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid parameter ID format")

        success = await parameter_service.delete_parameter(parameter_id)

        if not success:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Parameter not found")

        return {"success": True, "message": "Parameter deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Delete parameter error: {e}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))
